src/lib.rs

```diff
@@ -312,21 +312,26 @@ impl<T, F, S> ScopeGuard<T, F, S>
     ///     if conditional() {
     ///         // a condition maybe makes us decide to
     ///         // “defuse” the guard and get back its inner parts
-    ///         let (value, dropfn) = ScopeGuard::into_inner(guard);
+    ///         let value = ScopeGuard::into_inner(guard);
     ///     } else {
     ///         // guard still exists in this branch
     ///     }
     /// }
     /// ```
     #[inline]
-    pub fn into_inner(guard: Self) -> (T, F) {
+    pub fn into_inner(guard: Self) -> T {
         // Cannot pattern match out of Drop-implementing types, so
-        // ptr::read the types to return and forget the source.
+        // ptr::read the value and drop only the the closure.
         unsafe {
-            let value = ptr::read(&*guard.value);
-            let dropfn = ptr::read(&*guard.dropfn);
-            mem::forget(guard);
-            (value, dropfn)
+            // inhibit the guard's Drop impl
+            let mut defused = ManuallyDrop::new(guard);
+            // assign the value to a local variable before dropping the
+            // closure, to ensure that the value is dropped if the closure's
+            // Drop impl panicks.
+            let value = ptr::read(&*defused.value);
+            // drop the closure (and any captured variables consumed by it)
+            ptr::drop_in_place(&mut *defused.dropfn);
```

Comment:
Can't we just invert the order of the drop and then we don't need the extra manually drop?

Reply:
Hm ok maybe I'd instead read both vars like before and let the closure drop. It doesn't need to be more complex I think

Reply:
While only changing the return line would work, I don't like how correctness would then rely on the exact placement of the implicit drop. It would also need comments to explain that the closure must be read and assigned to a local variable. The new code makes that drop explicit.

Reply:
The diff would be like this:

```diff
     #[inline]
-    pub fn into_inner(guard: Self) -> (T, F) {
+    pub fn into_inner(guard: Self) -> T {
         // Cannot pattern match out of Drop-implementing types, so
         // ptr::read the types to return and forget the source.
         unsafe {
             let value = ptr::read(&*guard.value);
             let dropfn = ptr::read(&*guard.dropfn);
             mem::forget(guard);
-            (value, dropfn)
+            value
         }
     }
 }
```

it looks like the simplest solution to me.

Reply:
Yes, but then it's tempting to later remove `let _dropfn = ptr::read(&*guard.dropfn);` or change it to `let _ = ptr::read(&*guard.dropfn);`

Reply:
OK

Reply:
I apparently forgot to actually add the changes when I amended yesterday :smile:

Pushed now.